USca/USca_ReportManager/Controls/ReportAlarmsByDateRange.xaml.cs

```diff
@@ -0,0 +1,69 @@
+﻿using System;
+using System.Collections.ObjectModel;
+using System.Linq;
+using System.Windows;
+using System.Windows.Controls;
+using System.Windows.Input;
+using USca_ReportManager.Util;
+
+namespace USca_ReportManager.Controls
+{
+    /// <summary>
+    /// Interaction logic for ReportAlarmsByDateRange.xaml
+    /// </summary>
+    public partial class ReportAlarmsByDateRange : UserControl
+    {
+        public ObservableCollection<AlarmLogDTO> AlarmLogs { get; set; } = new();
+        private AlarmLogService _alarmLogService = new();  // TODO: do a singleton?
+        private DateTime? _startTime = null;
+        private DateTime? _endTime = null;
+
+        public ReportAlarmsByDateRange()
+        {
+            InitializeComponent();
+        }
+
+        private void Calendar_SelectedDatesChanged(object sender, SelectionChangedEventArgs e)
+        {
+            Mouse.Capture(null);
+            Calendar calendar = (Calendar)sender;
+            try
+            {
+                _startTime = calendar.SelectedDates[0];
+                _endTime = calendar.SelectedDates.Last();
+            }
+            catch (ArgumentOutOfRangeException)
+            {
+                return;
+            }
+        }
+
+        private async void BtnSearch_Click(object sender, RoutedEventArgs e)
+        {
+            if (_startTime == null || _endTime == null)
+            {
+                MessageBox.Show("Must select a date range!", "Failure", MessageBoxButton.OK);
+                return;
+            }
+            if (_startTime >= _endTime)
+            {
+                MessageBox.Show("Start must come before end!", "Failure", MessageBoxButton.OK);
+                return;
+            }
+            try
+            {
+                var res = await _alarmLogService.GetByDateRange((DateTime) _startTime, (DateTime) _endTime);
```

Comment:
Why are we casting `_startTime` and `_endTime`?

Reply:
Because they are nullable before

Reply:
You do a null check in the code above, after which point they get promoted to full `DateTime`s. Cast is unnecessary.

Reply:
VS complained, will check again

Reply:
Yup, it complained again without casting. It doesn't complain for you?

Reply:
It does complain actually, my bad. Shame we have to do casts.

Reply:
Only the best language doesn't complain here: JavaScript (TypeScript actually)